teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,279 @@
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+- [Out-of-scope](#out-of-scope)
+- [Future Work](#future-work)
+<!-- /toc -->
+
+## Summary
+This proposal outlines what information is required in the [provenance](#provenance-for-executed-taskrun)to reproduce complete build instructions for taskruns. It also suggests where in the `SLSA v0.2` [provenance](https://slsa.dev/provenance/v0.2) to store this information. 
+ 
+## Motivation
+
+[Tekton Chains will report provenance as reproducible if a specific annotation is included in the TaskRun](https://docs.google.com/document/d/12yYP0M_zz5Tc9ftGnLwhDZ_tdZBq4mEod70ngWtHmLs/edit#bookmark=id.1sm6isih1d0n) and it includes Task steps and parameter values in the provenance it generates, however there is a great deal of Task spec contents it ignores (e.g. sidecar definitions) and runtime information it does not include (e.g. workspaces provided at runtime). 
+
+For example the following provenance snippet was generated by Tekton Chains ([v0.8.0](https://github.com/tektoncd/chains/releases/tag/v0.8.0)) from [this chains example Task](https://github.com/tektoncd/chains/blob/main/examples/kaniko/kaniko.yaml) - it contains step information (including the shas of the images run) but is missing other information such as the workspaces and the volumes that backed them:
+
+```yaml
+invocation:
+  configSource: {}
+  parameters:
+    BUILDER_IMAGE: gcr.io/kaniko-project/executor:v1.5.1@sha256:c6166717f7fe0b7da44908c986137ecfeab21f31ec3992f6e128fff8a94be8a5
+    CONTEXT: ./
+    DOCKERFILE: ./Dockerfile
+    EXTRA_ARGS: '[]'
+    IMAGE: '{string us.gcr.io/christiewilson-catfactory/kaniko-chains []}'
+buildConfig:
+  steps:
+    - entryPoint: |
+        set -e
+        echo "FROM alpine@sha256:69e70a79f2d41ab5d637de98c1e0b055206ba40a8145e7bddb55ccc04e13cf8f" | tee $(params.DOCKERFILE)
+      arguments: null
+      environment:
+        container: add-dockerfile
+        image: docker-pullable://bash@sha256:fc742d0c3d9d8f5fb2681062398c04b710cd08c46dac1a8f0a5515687018acb9
+      annotations: null
+    - entryPoint: ""
+      arguments:
+        - $(params.EXTRA_ARGS)
+        - --dockerfile=$(params.DOCKERFILE)
+        - --context=$(workspaces.source.path)/$(params.CONTEXT)
+        - --destination=$(params.IMAGE)
+        - --digest-file=$(results.IMAGE_DIGEST.path)
+      environment:
+        container: build-and-push
+        image: docker-pullable://gcr.io/kaniko-project/executor@sha256:68bb272f681f691254acfbdcef00962f22efe2f0c1e287e6a837b0abe07fb94b
+      annotations: null
+    - entryPoint: |
+        set -e
+        echo $(params.IMAGE) | tee $(results.IMAGE_URL.path)
+      arguments: null
+      environment:
+        container: write-url
+        image: docker-pullable://bash@sha256:fc742d0c3d9d8f5fb2681062398c04b710cd08c46dac1a8f0a5515687018acb9
+      annotations: null
+```
+
+## Requirements
```

Comment:
Have you considered capturing the status of network isolation in this proposal?

Reply:
I did not. May be we can capture this in the out-of-scope section of this proposal since we may need this for the **Full** level of reproducibility (see last two sections).

Reply:
I summarized the point about Hermetic builds. For now, it is in the out-of-scope section since the goal of this TEP is to capture information that is missing in the provenance and is absolutely needed for reproducibility.